### ISSUE TITLE
Locally abstracting over async calls

### ISSUE THREAD

**Issue by nomeata** [dfinity/motoko#680]
An unrelated discussion (“Could secure randomness be an asynchronous rather than synchronous system feature”) made me think of the following problem with our language design. I’ll use that as an example. 

Consider we have a `getRand : shared () -> Word64`. Obviously, I can use this to get randomness of various sorts:
```
async {
  let tmp1 = await getRand();
  let randomBool = if (tmp1 % 2 = 0) true else false;
  let tmp2 = await getRand();
  let randomWord32 = natToWord32(word64ToNat(tmp2));
}
```
But really, we want to be able to provide a library of various `getRandX()` calls to abstract out these patterns.

But we can't at the moment:
* ```
  module Rand {
    func getRandBool() : Bool {
      let tmp = await getRand();
      return (if (tmp1 % 2 = 0) true else false);
    } 
  } : module { getRandBool : () -> Bool }
  ```
  doesn't work because we can’t use `await` in this context, and because the type would be a lie.
* ```
  module Rand {
    func getRandBool() : async Bool {
      async {
        let tmp = await getRand();
        return (if (tmp1 % 2 = 0) true else false);
      }
    } 
  } : module { getRandBool : () -> async Bool }
  ```
  works, but the free-standing `async` block would cause an extra shared function call whenver this function is used
* ```
  module Rand {
    shared func getRandBool() : async Bool {
      let tmp = await getRand();
      return (if (tmp1 % 2 = 0) true else false);
    } 
  } : module { getRandBool : shared () -> async Bool }
  ```
  has the same problem.

So it seems that we need a way to define a function that
 * has it’s body in `async` context (so it can do `await`)
 * can itself only be used in `async` context
 * in other words: returns its result by invoking a continuation
 * but does not on its own cause an async all
 * is (probably) different from the `shared A -> B` type (as it is not transferable, cannot be used as a canister method)

In other words:
 * We have `Local`/`Return`: Normal function
 * We have  `Shared`/`Promise`: Normal shared function
 * We have  `Shared`/`Return`: One-Shot shared function (but the `Return` here only refers to control flow, not to any return values, of which there are not any, also see #661).
 * This asks for: `Local`/`Promise`: Local functions that take a continuation.

Original discussion on Slack: https://dfinity.slack.com/archives/CCK4B0XE3/p1569065543185500
